### DISCUSSION TITLE
Bug? Specify that a TypeVar supports the "-" operator among its values

### DISCUSSION THREAD

**Discussion by ayhon** [microsoft/pyright#5641] [Q&A] [answered]
I asked about specifying that a `TypeVar` supports subtraction [on StackOverflow](https://stackoverflow.com/q/76821158/13561543), and it seems like the solution seems to be the following combination of `Protocol` and `Self`
```python
from typing import TypeVar, Generic, Protocol, Self

class SupportsSub(Protocol):
    @abstractmethod
    def __sub__(self, other: Self) -> Self: pass
```
with this I can type the following class
```python
T = TypeVar('T', bound=SupportsSub)
@dataclass
class Consumption(Generic[T]):
    first: T
    second: T
    def difference(self) -> T:
        return self.second - self.first
class Other:
    water: Consumption[float]
```

However, this is rejected by `pyright` but [accepted by `mypy`](https://mypy-play.net/?mypy=latest&python=3.11&flags=strict&gist=306d7aaf0b3d80aa406689eb27ead302). I wanted to check if this is a bug before opening an issue

**Comment by debonte** [comment, marked as answer]
This is because the `other` parameter in `SupportsSub` can be passed as either a positional or keyword parameter, whereas the parameter to `__sub__` on `float` is positional-only due to the `__` prefix. If you make your `__sub__` parameter positional-only, the error goes away:

```python
def __sub__(self, other: Self, /) -> Self:
```

For others, the error seen was:

```text
Type "float" cannot be assigned to type variable "T@Consumption"
  Type "float" is incompatible with bound type "SupportsSub" for type variable "T@Consumption"
    "float" is incompatible with protocol "SupportsSub"
      "__sub__" is an incompatible type
        Type "(__value: float, /) -> float" cannot be assigned to type "(other: float) -> float"
          Position-only parameter mismatch; expected 1 but received 0
```

**Comment by ayhon** [comment]
I see, thanks a lot!
